desc/objectives/_coils.py

```diff
@@ -708,6 +716,362 @@ def compute(self, params, constants=None):
+        Collocation grid used to discritize each coil. Default = LinearGrid(N=16)
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _units = "(m)"
+    _print_value_fmt = "Minimum coil-coil distance: {:10.3e} "
+
+    def __init__(
+        self,
+        coils,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name="coil-coil minimum distance",
+    ):
+        if target is None and bounds is None:
+            bounds = (1, np.inf)
+        self._grid = grid
+        super().__init__(
+            things=coils,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        coilset = self.things[0]
+        grid = self._grid or LinearGrid(N=16)
+
+        self._dim_f = coilset.num_coils
+        self._constants = {"coilset": coilset, "grid": grid, "quad_weights": 1.0}
+
+        if self._normalize:
+            coils = tree_leaves(coilset, is_leaf=lambda x: not hasattr(x, "__len__"))
+            scales = [compute_scaling_factors(coil)["a"] for coil in coils]
+            self._normalization = np.mean(scales)  # mean length of coils
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """Compute minimum distances between coils.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of coilset degrees of freedom, eg CoilSet.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc.
+            Defaults to self._constants.
+
+        Returns
+        -------
+        f : array of floats
+            Minimum distance to another coil for each coil in the coilset.
+
+        """
+        if constants is None:
+            constants = self.constants
+        pts = constants["coilset"]._compute_position(
+            params=params, grid=constants["grid"]
+        )
+
+        def body(k):
+            # dist btwn all pts; shape(ncoils,num_nodes,num_nodes)
+            dist = safenorm(pts[k][None, :, None] - pts[:, None, :], axis=-1)
+            # exclude distances between points on the same coil
+            mask = jnp.ones(self.dim_f).at[k].set(0)[:, None, None]
+            return jnp.min(dist, where=mask, initial=jnp.inf)
+
+        min_dist_per_coil = fori_loop(
```

Comment:
does this need to be a loop or can we do vmap + reduction? It might be good to have both options with a flag like for the free boundary objective? or a batched vmap or something?

Reply:
make issue